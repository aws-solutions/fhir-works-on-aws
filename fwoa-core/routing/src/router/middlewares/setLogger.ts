/* eslint-disable @typescript-eslint/no-unused-vars */
/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 *
 */

import express, { json } from 'express';
import _ from 'lodash';
import { v4 } from 'uuid';
import getComponentLogger, { getEncryptLogger } from '../../loggerBuilder';
import {
  captureFhirUserParts,
  capturePathParts,
  capturePathParametersProxyParts
} from '../../regExpressions';

// define metadata,who,what,when,where,how,requestOther,responseOther 5 categoires in logger

export interface metaDataAttribute {
  uid: string;
  timestamp: string;
  category: string;
  encryptedFields?: string[];
  payLoadToEncrypt?: { [key: string]: string };
  encryptedPayLoad?: string;
}

export interface whoAttribute {
  userIdentity: {
    sub: string;
    fhirUser?: string;
    'cognito:groups'?: string[];
    'cognito:username'?: string;
    'custom:tenantId'?: string;
  };
  apiKeyId: string;
}

export interface whatAttribute {
  userIdentity: {
    scopes?: string[];
    usableScopes?: string[];
  };
  requestContext: { path: string; httpMethod: string };
  apiGateway: {
    event: {
      httpMethod: string;
      queryStringParameters?: { [key: string]: string };
      pathParameters?: {
        proxy: string;
      };
    };
  };
}
export interface whenAttribute {
  requestContext: { requestTimeEpoch: string };
}

export interface whereAttribute {
  apiGateway: {
    Context: {
      logGroupName: string;
      logStreamName: string;
    };
  };
  'User-agent': string;
  requestContext: {
    domainName: string;
    identity: {
      sourceIp: string;
    };
  };
}

export interface howAttribute {
  apiGateway: {
    context: { awsRequestId: string };
  };
  userIdentity: { jti: string };
}
export interface requestOtherAttribute {
  requestContext: {
    stage: string;
  };
}

export interface responseOtherAttribute {
  userIdentity: {
    'launch-response-patient'?: string;
    iss: string; // issuer: identify who created and signed the access token
    aud: string; // The dev URL of the fhir works that you're trying to access using the JWT to authenticate
    scp?: string[]; // The scopes from the original request
    iat: string; // When the token was issued
    exp: string; // The token expiration time
    'auth-time': string; // when the authentication occurs
  };
}

export interface setEntireLogger {
  logMetadata: metaDataAttribute;
  who: whoAttribute;
  what: whatAttribute;
  when: whenAttribute;
  where: whereAttribute;
  how: howAttribute;
  requestOther: requestOtherAttribute;
  responseOther: responseOtherAttribute;
}

function preprocessFieldsToEncrypted(fieldsToEncrypt: string[], logging: setEntireLogger): setEntireLogger {
  // mark encrypted string for selected elements
  const fieldsToEncryptedObject: { [key: string]: string } = {};
  const loggingPreprocessed = { ...logging };
  const validFieldsToEncrypt: string[] = [];
  const stringToReplace = 'encrypted';
  fieldsToEncrypt.forEach((selectedField) => {
    let markedContent: string | undefined | null;
    // eslint-disable-next-line default-case
    const contentInField = _.get(logging, selectedField);
    if (contentInField) {
      switch (selectedField) {
        case 'who.userIdentity.sub': {
          markedContent = stringToReplace;
          break;
        }
        // {fhirUser:https://example.execute-api.us-east-1.amazonaws.com/dev/Practitioner/00000000-0000-0000-0000-000000000000}
        // {fhirUser:https://example.execute-api.us-east-1.amazonaws.com/dev/Patient/00000000-0000-0000-0000-000000000000}
        // match groups = ["https://example.execute-api.us-east-1.amazonaws.com/dev/Patient/00000000-0000-0000-0000-000000000000", "https://example.execute-api.us-east-1.amazonaws.com/dev/Patient/", "00000000-0000-0000-0000-000000000000", undefined]
        case 'who.userIdentity.fhirUser': {
          const actualMatch = contentInField.match(captureFhirUserParts);
          if (actualMatch && actualMatch[2]) {
            markedContent = actualMatch[1] + stringToReplace;
          } else {
            markedContent = contentInField;
          }
          break;
        }
        // queryStringParameters: {name: 'FakeName'}
        // queryStringParameters: null
        case 'what.apiGateway.event.queryStringParameters': {
          markedContent = stringToReplace;
          break;
        }
        // {path: '/dev/Patient/00000000-0000-0000-0000-000000000000'}
        // {path: '/dev/Practitioner'}
        // match groups = ["/dev/Patient/00000000-0000-0000-0000-000000000000", "/dev/Patient/", "00000000-0000-0000-0000-000000000000"]
        case 'what.requestContext.path': {
          const actualMatch = contentInField.match(capturePathParts);
          if (actualMatch && actualMatch[2]) {
            markedContent = actualMatch[1] + stringToReplace;
          } else {
            markedContent = contentInField;
          }
          break;
        }
        // {pathParameters:{ proxy: 'patient/00000000-0000-0000-0000-000000000000' }}
        // {pathParameters:{}}
        // match groups =["Patient/00000000-0000-0000-0000-000000000000", "Patient/", "00000000-0000-0000-0000-000000000000"]
        case 'what.apiGateway.event.pathParameters.proxy': {
          const actualMatch = contentInField.match(capturePathParametersProxyParts);
          if (actualMatch && actualMatch[2]) {
            markedContent = actualMatch[1] + stringToReplace;
          } else {
            markedContent = contentInField;
          }
          break;
        }
        // {sourceIp: '0.0.0.0'}
        case 'where.requestContext.identity.sourceIp': {
          markedContent = stringToReplace;
          break;
        }
        // {launch-response-patient:'https://example.execute-api.us-east-1.amazonaws.com/dev/Patient/22222222-2222-2222-2222-222222222222'}
        // {launch-response-patient:'Patient/22222222-2222-2222-2222-222222222222'}
        // {launch-response-patient:''}
        case 'responseOther.userIdentity.launch-response-patient': {
          const actualMatchCase1 = contentInField.match(captureFhirUserParts);
          const actualMatchCase2 = contentInField.match(capturePathParametersProxyParts);
          if (actualMatchCase1 && actualMatchCase1[2]) {
            markedContent = actualMatchCase1[1] + stringToReplace;
          } else if (actualMatchCase2 && actualMatchCase2[2]) {
            markedContent = actualMatchCase2[1] + stringToReplace;
          } else {
            markedContent = contentInField;
          }
          break;
        }
        default:
          throw new Error('Field to encrypt must be the existing cases');
      }
      if (markedContent !== contentInField) {
        _.set(fieldsToEncryptedObject, selectedField, contentInField);
        _.set(loggingPreprocessed, selectedField, markedContent);
        validFieldsToEncrypt.push(selectedField);
      }
    }
  });
  if (Object.keys(fieldsToEncryptedObject).length !== 0) {
    _.set(loggingPreprocessed, 'logMetadata.encryptedFields', validFieldsToEncrypt);
    _.set(loggingPreprocessed, 'logMetadata.payLoadToEncrypt', fieldsToEncryptedObject);
  }
  return loggingPreprocessed;
}

// define logger middleware
export const setLoggerMiddleware = async (
  req: express.Request,
  res: express.Response,
  next: express.NextFunction
) => {
  try {
    const response = res as any;
    const request = req as any;
    const requestTimeEpochDate = new Date(request.requestContext.requestTimeEpoch);
    const iatDate = new Date(res.locals.userIdentity.iat * 1000);
    const expDate = new Date(res.locals.userIdentity.exp * 1000);
    const authTimeDate = new Date(res.locals.userIdentity.auth_time * 1000);
    const timestampDate = new Date();
    let loggerDesign: setEntireLogger = {
      logMetadata: {
        uid: v4(),
        timestamp: timestampDate.toISOString(),
        category: 'Audit Event'
      },
      who: {
        userIdentity: {
          sub: response.locals.userIdentity.sub,
          fhirUser: response.locals.userIdentity.fhirUser,
          'cognito:groups': response.locals.userIdentity['cognito:groups'],
          'cognito:username': response.locals.userIdentity['cognito:username'],
          'custom:tenantId': response.locals.userIdentity['custom:tenantId']
        },
        apiKeyId: request.requestContext.identity.apiKeyId
      },
      what: {
        userIdentity: (({ scopes, usableScopes }) => ({ scopes, usableScopes }))(
          response.locals.userIdentity
        ),
        requestContext: (({ path, httpMethod }) => ({ path, httpMethod }))(request.requestContext),
        apiGateway: {
          event: (({ httpMethod, queryStringParameters, pathParameters }) => ({
            httpMethod,
            queryStringParameters,
            pathParameters
          }))(request.apiGateway.event)
        }
      },
      when: {
        requestContext: { requestTimeEpoch: requestTimeEpochDate.toISOString() }
      },
      where: {
        apiGateway: {
          Context: (({ logGroupName, logStreamName }) => ({ logGroupName, logStreamName }))(
            request.apiGateway.context
          )
        },
        'User-agent': request.headers['user-agent'],
        requestContext: {
          domainName: request.requestContext.domainName,
          identity: (({ sourceIp }) => ({ sourceIp }))(request.requestContext.identity)
        }
      },

      how: {
        apiGateway: { context: (({ awsRequestId }) => ({ awsRequestId }))(request.apiGateway.context) },
        userIdentity: (({ jti }) => ({ jti }))(response.locals.userIdentity)
      },
      requestOther: {
        requestContext: {
          stage: request.requestContext.stage
        }
      },
      responseOther: {
        userIdentity: {
          ...{ 'launch-response-patient': response.locals.userIdentity.launch_response_patient },
          ...(({ iss, aud, scp }) => ({
            iss,
            aud,
            scp
          }))(response.locals.userIdentity),
          iat: iatDate.toISOString(),
          exp: expDate.toISOString(),
          'auth-time': authTimeDate.toISOString()
        }
      }
    };

    loggerDesign = preprocessFieldsToEncrypted(
      [
        'who.userIdentity.sub',
        'who.userIdentity.fhirUser',
        'what.apiGateway.event.queryStringParameters',
        'what.requestContext.path',
        'what.apiGateway.event.pathParameters.proxy',
        'where.requestContext.identity.sourceIp',
        'responseOther.userIdentity.launch-response-patient'
      ],
      loggerDesign
    );
    const encryptedField = 'logMetadata.payLoadToEncrypt';
    const logger = getEncryptLogger({ encryptedField });
    await logger.error(loggerDesign);

    next();
  } catch (e) {
    next(e);
  }
};
