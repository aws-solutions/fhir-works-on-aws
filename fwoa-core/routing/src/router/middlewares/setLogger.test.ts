const Logger = {
  error: jest.fn()
};
const encryptLogger = {
  error: jest.fn()
};

jest.mock('../../loggerBuilder', () => {
  const originalModule = jest.requireActual('../../loggerBuilder');
  return {
    __esModule: true,
    ...originalModule,
    default: jest.fn(() => Logger),
    getEncryptLogger: jest.fn(() => encryptLogger)
  };
});

jest.mock('uuid', () => ({ v4: () => '00000000-0000-0000-0000-000000000000' }));

jest.useFakeTimers().setSystemTime(new Date('2020-01-01'));
import express from 'express';
import { setLoggerMiddleware } from './setLogger';

describe('createLoggerMiddleware', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  describe('test encryption', () => {
    test('happy case', async () => {
      //BUILD
      process.env.ENABLE_LOGGING_MIDDLEWARE_ENCRYPTION = 'true';
      const nextMock = jest.fn();
      const req = {
        requestContext: {
          requestTimeEpoch: 1673369044,
          identity: {
            apiKeyId: 'FakeApiKeyId',
            sourceIp: '0.0.0.0'
          },
          domainName: 'FakeDomainName',
          path: '/dev/Patient',
          httpMethod: 'GET',
          stage: 'dev'
        },
        apiGateway: {
          event: {
            httpMethod: 'GET',
            queryStringParameters: {
              name: 'FakeName'
            },
            pathParameters: { proxy: 'patient/00000000-0000-0000-0000-000000000000' }
          },
          context: {
            logGroupName: 'FakeLogGroupName',
            logStreamName: 'FakeLogStreamName',
            awsRequestId: '11111111-1111-1111-1111-111111111111'
          }
        },
        headers: {
          'user-agent': 'FakeUserAgent'
        }
      } as unknown as express.Request;
      const res = {
        locals: {
          userIdentity: {
            sub: 'example@amazon.com',
            fhirUser:
              'https://example.execute-api.us-east-1.amazonaws.com/dev/Practitioner/00000000-0000-0000-0000-000000000000',
            scopes: ['FakeScope1', 'FakeScope2', 'FakeScope3'],
            usableScopes: ['FakeScope2', 'FakeScope3'],
            launch_response_patient:
              'https://example.execute-api.us-east-1.amazonaws.com/dev/Patient/22222222-2222-2222-2222-222222222222',
            jti: 'AT.FakeJTIValue',
            iss: 'FakeIssuer',
            aud: 'FakeAudience',
            iat: 167338044,
            exp: 167339044,
            auth_time: 1673386822,
            scp: ['FakeScope1', 'FakeScope2', 'FakeScope3']
          }
        }
      } as unknown as express.Response;

      //OPERATE
      await setLoggerMiddleware(req, res, nextMock);

      //CHECK
      expect(nextMock).toBeCalledTimes(1);
      expect(encryptLogger.error).toMatchSnapshot();
    });

    test('test null value in a field for encryption case', async () => {
      //BUILD
      process.env.ENABLE_LOGGING_MIDDLEWARE_ENCRYPTION = 'true';
      const nextMock = jest.fn();
      const req = {
        requestContext: {
          requestTimeEpoch: 167337044,
          identity: {
            apiKeyId: 'FakeApiKeyId',
            sourceIp: '0.0.0.0'
          },
          domainName: 'FakeDomainName',
          path: '/dev/Patient',
          httpMethod: 'GET',
          stage: 'dev'
        },
        apiGateway: {
          event: {
            httpMethod: 'GET',
            queryStringParameters: null,
            pathParameters: { proxy: 'patient/00000000-0000-0000-0000-000000000000' }
          },
          context: {
            logGroupName: 'FakeLogGroupName',
            logStreamName: 'FakeLogStreamName',
            awsRequestId: '11111111-1111-1111-1111-111111111111'
          }
        },
        headers: {
          'user-agent': 'FakeUserAgent'
        }
      } as unknown as express.Request;
      const res = {
        locals: {
          userIdentity: {
            sub: 'example@amazon.com',
            fhirUser:
              'https://example.execute-api.us-east-1.amazonaws.com/dev/Practitioner/00000000-0000-0000-0000-000000000000',
            scopes: ['FakeScope1', 'FakeScope2', 'FakeScope3'],
            usableScopes: ['FakeScope2', 'FakeScope3'],
            launch_response_patient: '',
            jti: 'AT.FakeJTIValue',
            iss: 'FakeIssuer',
            aud: 'FakeAudience',
            iat: 167338044,
            exp: 167339044,
            auth_time: 1673386822,
            scp: ['FakeScope1', 'FakeScope2', 'FakeScope3']
          }
        }
      } as unknown as express.Response;
      //OPERATE
      await setLoggerMiddleware(req, res, nextMock);

      //CHECK
      expect(nextMock).toBeCalledTimes(1);
      expect(encryptLogger.error).toMatchSnapshot();
    });
    test('test reference launch context value for encryption case', async () => {
      //BUILD
      process.env.ENABLE_LOGGING_MIDDLEWARE_ENCRYPTION = 'true';
      const nextMock = jest.fn();
      const req = {
        requestContext: {
          requestTimeEpoch: 167337044,
          identity: {
            apiKeyId: 'FakeApiKeyId',
            sourceIp: '0.0.0.0'
          },
          domainName: 'FakeDomainName',
          path: '/dev/Patient',
          httpMethod: 'GET',
          stage: 'dev'
        },
        apiGateway: {
          event: {
            httpMethod: 'GET',
            queryStringParameters: null,
            pathParameters: { proxy: 'patient/00000000-0000-0000-0000-000000000000' }
          },
          context: {
            logGroupName: 'FakeLogGroupName',
            logStreamName: 'FakeLogStreamName',
            awsRequestId: '11111111-1111-1111-1111-111111111111'
          }
        },
        headers: {
          'user-agent': 'FakeUserAgent'
        }
      } as unknown as express.Request;
      const res = {
        locals: {
          userIdentity: {
            sub: 'example@amazon.com',
            fhirUser:
              'https://example.execute-api.us-east-1.amazonaws.com/dev/Practitioner/00000000-0000-0000-0000-000000000000',
            scopes: ['FakeScope1', 'FakeScope2', 'FakeScope3'],
            usableScopes: ['FakeScope2', 'FakeScope3'],
            launch_response_patient: 'Patient/22222222-2222-2222-2222-222222222222',
            jti: 'AT.FakeJTIValue',
            iss: 'FakeIssuer',
            aud: 'FakeAudience',
            iat: 167338044,
            exp: 167339044,
            auth_time: 1673386822,
            scp: ['FakeScope1', 'FakeScope2', 'FakeScope3']
          }
        }
      } as unknown as express.Response;
      //OPERATE
      await setLoggerMiddleware(req, res, nextMock);

      //CHECK
      expect(nextMock).toBeCalledTimes(1);
      expect(encryptLogger.error).toMatchSnapshot();
    });
  });
  test('test no encrypt case', async () => {
    //BUILD
    process.env.ENABLE_LOGGING_MIDDLEWARE_ENCRYPTION = 'false';
    const nextMock = jest.fn();
    const req = {
      requestContext: {
        requestTimeEpoch: 167337044,
        identity: {
          apiKeyId: 'FakeApiKeyId',
          sourceIp: '0.0.0.0'
        },
        domainName: 'FakeDomainName',
        path: '/dev/Patient',
        httpMethod: 'GET',
        stage: 'dev'
      },
      apiGateway: {
        event: {
          httpMethod: 'GET',
          queryStringParameters: {
            name: 'FakeName'
          },
          pathParameters: { proxy: 'patient/00000000-0000-0000-0000-000000000000' }
        },
        context: {
          logGroupName: 'FakeLogGroupName',
          logStreamName: 'FakeLogStreamName',
          awsRequestId: '11111111-1111-1111-1111-111111111111'
        }
      },
      headers: {
        'user-agent': 'FakeUserAgent'
      }
    } as unknown as express.Request;
    const res = {
      locals: {
        userIdentity: {
          sub: 'example@amazon.com',
          fhirUser:
            'https://example.execute-api.us-east-1.amazonaws.com/dev/Practitioner/00000000-0000-0000-0000-000000000000',
          scopes: ['FakeScope1', 'FakeScope2', 'FakeScope3'],
          usableScopes: ['FakeScope2', 'FakeScope3'],
          launch_response_patient:
            'https://example.execute-api.us-east-1.amazonaws.com/dev/Practitioner/00000000-0000-0000-0000-000000000000',
          jti: 'AT.FakeJTIValue',
          iss: 'FakeIssuer',
          aud: 'FakeAudience',
          iat: 167338044,
          exp: 167339044,
          auth_time: 1673386822,
          scp: ['FakeScope1', 'FakeScope2', 'FakeScope3']
        }
      }
    } as unknown as express.Response;
    //OPERATE
    await setLoggerMiddleware(req, res, nextMock);

    //CHECK
    expect(nextMock).toBeCalledTimes(1);
    expect(Logger.error).toMatchSnapshot();
  });
});
