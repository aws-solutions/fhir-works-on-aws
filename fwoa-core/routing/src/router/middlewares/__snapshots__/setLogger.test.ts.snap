// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createLoggerMiddleware test encryption happy case 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "how": Object {
          "apiGateway": Object {
            "context": Object {
              "awsRequestId": "11111111-1111-1111-1111-111111111111",
            },
          },
          "userIdentity": Object {
            "jti": "AT.FakeJTIValue",
          },
        },
        "logMetadata": Object {
          "category": "Audit Event",
          "encryptedFields": Array [
            "who.userIdentity.sub",
            "who.userIdentity.fhirUser",
            "what.apiGateway.event.queryStringParameters",
            "what.apiGateway.event.pathParameters.proxy",
            "where.requestContext.identity.sourceIp",
            "responseOther.userIdentity.launch-response-patient",
          ],
          "payLoadToEncrypt": Object {
            "responseOther": Object {
              "userIdentity": Object {
                "launch-response-patient": "https://example.execute-api.us-east-1.amazonaws.com/dev/Patient/22222222-2222-2222-2222-222222222222",
              },
            },
            "what": Object {
              "apiGateway": Object {
                "event": Object {
                  "pathParameters": Object {
                    "proxy": "patient/00000000-0000-0000-0000-000000000000",
                  },
                  "queryStringParameters": Object {
                    "name": "FakeName",
                  },
                },
              },
            },
            "where": Object {
              "requestContext": Object {
                "identity": Object {
                  "sourceIp": "0.0.0.0",
                },
              },
            },
            "who": Object {
              "userIdentity": Object {
                "fhirUser": "https://example.execute-api.us-east-1.amazonaws.com/dev/Practitioner/00000000-0000-0000-0000-000000000000",
                "sub": "example@amazon.com",
              },
            },
          },
          "timestamp": "2020-01-01T00:00:00.000Z",
          "uid": "00000000-0000-0000-0000-000000000000",
        },
        "requestOther": Object {
          "requestContext": Object {
            "stage": "dev",
          },
        },
        "responseOther": Object {
          "userIdentity": Object {
            "aud": "FakeAudience",
            "auth-time": "2023-01-10T21:40:22.000Z",
            "exp": "1975-04-21T19:04:04.000Z",
            "iat": "1975-04-21T18:47:24.000Z",
            "iss": "FakeIssuer",
            "launch-response-patient": "https://example.execute-api.us-east-1.amazonaws.com/dev/Patient/encrypted",
            "scp": Array [
              "FakeScope1",
              "FakeScope2",
              "FakeScope3",
            ],
          },
        },
        "what": Object {
          "apiGateway": Object {
            "event": Object {
              "httpMethod": "GET",
              "pathParameters": Object {
                "proxy": "patient/encrypted",
              },
              "queryStringParameters": "encrypted",
            },
          },
          "requestContext": Object {
            "httpMethod": "GET",
            "path": "/dev/Patient",
          },
          "userIdentity": Object {
            "scopes": Array [
              "FakeScope1",
              "FakeScope2",
              "FakeScope3",
            ],
            "usableScopes": Array [
              "FakeScope2",
              "FakeScope3",
            ],
          },
        },
        "when": Object {
          "requestContext": Object {
            "requestTimeEpoch": "1970-01-20T08:49:29.044Z",
          },
        },
        "where": Object {
          "User-agent": "FakeUserAgent",
          "apiGateway": Object {
            "Context": Object {
              "logGroupName": "FakeLogGroupName",
              "logStreamName": "FakeLogStreamName",
            },
          },
          "requestContext": Object {
            "domainName": "FakeDomainName",
            "identity": Object {
              "sourceIp": "encrypted",
            },
          },
        },
        "who": Object {
          "apiKeyId": "FakeApiKeyId",
          "userIdentity": Object {
            "cognito:groups": undefined,
            "cognito:username": undefined,
            "custom:tenantId": undefined,
            "fhirUser": "https://example.execute-api.us-east-1.amazonaws.com/dev/Practitioner/encrypted",
            "sub": "encrypted",
          },
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`createLoggerMiddleware test encryption test null value in a field for encryption case 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "how": Object {
          "apiGateway": Object {
            "context": Object {
              "awsRequestId": "11111111-1111-1111-1111-111111111111",
            },
          },
          "userIdentity": Object {
            "jti": "AT.FakeJTIValue",
          },
        },
        "logMetadata": Object {
          "category": "Audit Event",
          "encryptedFields": Array [
            "who.userIdentity.sub",
            "who.userIdentity.fhirUser",
            "what.apiGateway.event.pathParameters.proxy",
            "where.requestContext.identity.sourceIp",
          ],
          "payLoadToEncrypt": Object {
            "what": Object {
              "apiGateway": Object {
                "event": Object {
                  "pathParameters": Object {
                    "proxy": "patient/00000000-0000-0000-0000-000000000000",
                  },
                },
              },
            },
            "where": Object {
              "requestContext": Object {
                "identity": Object {
                  "sourceIp": "0.0.0.0",
                },
              },
            },
            "who": Object {
              "userIdentity": Object {
                "fhirUser": "https://example.execute-api.us-east-1.amazonaws.com/dev/Practitioner/00000000-0000-0000-0000-000000000000",
                "sub": "example@amazon.com",
              },
            },
          },
          "timestamp": "2020-01-01T00:00:00.000Z",
          "uid": "00000000-0000-0000-0000-000000000000",
        },
        "requestOther": Object {
          "requestContext": Object {
            "stage": "dev",
          },
        },
        "responseOther": Object {
          "userIdentity": Object {
            "aud": "FakeAudience",
            "auth-time": "2023-01-10T21:40:22.000Z",
            "exp": "1975-04-21T19:04:04.000Z",
            "iat": "1975-04-21T18:47:24.000Z",
            "iss": "FakeIssuer",
            "launch-response-patient": "",
            "scp": Array [
              "FakeScope1",
              "FakeScope2",
              "FakeScope3",
            ],
          },
        },
        "what": Object {
          "apiGateway": Object {
            "event": Object {
              "httpMethod": "GET",
              "pathParameters": Object {
                "proxy": "patient/encrypted",
              },
              "queryStringParameters": null,
            },
          },
          "requestContext": Object {
            "httpMethod": "GET",
            "path": "/dev/Patient",
          },
          "userIdentity": Object {
            "scopes": Array [
              "FakeScope1",
              "FakeScope2",
              "FakeScope3",
            ],
            "usableScopes": Array [
              "FakeScope2",
              "FakeScope3",
            ],
          },
        },
        "when": Object {
          "requestContext": Object {
            "requestTimeEpoch": "1970-01-02T22:28:57.044Z",
          },
        },
        "where": Object {
          "User-agent": "FakeUserAgent",
          "apiGateway": Object {
            "Context": Object {
              "logGroupName": "FakeLogGroupName",
              "logStreamName": "FakeLogStreamName",
            },
          },
          "requestContext": Object {
            "domainName": "FakeDomainName",
            "identity": Object {
              "sourceIp": "encrypted",
            },
          },
        },
        "who": Object {
          "apiKeyId": "FakeApiKeyId",
          "userIdentity": Object {
            "cognito:groups": undefined,
            "cognito:username": undefined,
            "custom:tenantId": undefined,
            "fhirUser": "https://example.execute-api.us-east-1.amazonaws.com/dev/Practitioner/encrypted",
            "sub": "encrypted",
          },
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`createLoggerMiddleware test encryption test reference launch context value for encryption case 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "how": Object {
          "apiGateway": Object {
            "context": Object {
              "awsRequestId": "11111111-1111-1111-1111-111111111111",
            },
          },
          "userIdentity": Object {
            "jti": "AT.FakeJTIValue",
          },
        },
        "logMetadata": Object {
          "category": "Audit Event",
          "encryptedFields": Array [
            "who.userIdentity.sub",
            "who.userIdentity.fhirUser",
            "what.apiGateway.event.pathParameters.proxy",
            "where.requestContext.identity.sourceIp",
            "responseOther.userIdentity.launch-response-patient",
          ],
          "payLoadToEncrypt": Object {
            "responseOther": Object {
              "userIdentity": Object {
                "launch-response-patient": "Patient/22222222-2222-2222-2222-222222222222",
              },
            },
            "what": Object {
              "apiGateway": Object {
                "event": Object {
                  "pathParameters": Object {
                    "proxy": "patient/00000000-0000-0000-0000-000000000000",
                  },
                },
              },
            },
            "where": Object {
              "requestContext": Object {
                "identity": Object {
                  "sourceIp": "0.0.0.0",
                },
              },
            },
            "who": Object {
              "userIdentity": Object {
                "fhirUser": "https://example.execute-api.us-east-1.amazonaws.com/dev/Practitioner/00000000-0000-0000-0000-000000000000",
                "sub": "example@amazon.com",
              },
            },
          },
          "timestamp": "2020-01-01T00:00:00.000Z",
          "uid": "00000000-0000-0000-0000-000000000000",
        },
        "requestOther": Object {
          "requestContext": Object {
            "stage": "dev",
          },
        },
        "responseOther": Object {
          "userIdentity": Object {
            "aud": "FakeAudience",
            "auth-time": "2023-01-10T21:40:22.000Z",
            "exp": "1975-04-21T19:04:04.000Z",
            "iat": "1975-04-21T18:47:24.000Z",
            "iss": "FakeIssuer",
            "launch-response-patient": "Patient/encrypted",
            "scp": Array [
              "FakeScope1",
              "FakeScope2",
              "FakeScope3",
            ],
          },
        },
        "what": Object {
          "apiGateway": Object {
            "event": Object {
              "httpMethod": "GET",
              "pathParameters": Object {
                "proxy": "patient/encrypted",
              },
              "queryStringParameters": null,
            },
          },
          "requestContext": Object {
            "httpMethod": "GET",
            "path": "/dev/Patient",
          },
          "userIdentity": Object {
            "scopes": Array [
              "FakeScope1",
              "FakeScope2",
              "FakeScope3",
            ],
            "usableScopes": Array [
              "FakeScope2",
              "FakeScope3",
            ],
          },
        },
        "when": Object {
          "requestContext": Object {
            "requestTimeEpoch": "1970-01-02T22:28:57.044Z",
          },
        },
        "where": Object {
          "User-agent": "FakeUserAgent",
          "apiGateway": Object {
            "Context": Object {
              "logGroupName": "FakeLogGroupName",
              "logStreamName": "FakeLogStreamName",
            },
          },
          "requestContext": Object {
            "domainName": "FakeDomainName",
            "identity": Object {
              "sourceIp": "encrypted",
            },
          },
        },
        "who": Object {
          "apiKeyId": "FakeApiKeyId",
          "userIdentity": Object {
            "cognito:groups": undefined,
            "cognito:username": undefined,
            "custom:tenantId": undefined,
            "fhirUser": "https://example.execute-api.us-east-1.amazonaws.com/dev/Practitioner/encrypted",
            "sub": "encrypted",
          },
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
